File: tools/registry.py
"""工具注册表"""
from typing import Dict, Any, Callable, Optional
from .base import Tool


class ToolRegistry:
    """工具注册表，管理所有可用工具"""
    
    def __init__(self):
        """初始化工具注册表"""
        self._tools: Dict[str, Tool] = {}
        self._functions: Dict[str, Dict[str, Any]] = {}
        print("[OK] 工具注册表初始化完成")
    
    def register_tool(self, tool: Tool) -> None:
        """
        注册工具类实例
        
        Args:
            tool: Tool实例
        """
        if not isinstance(tool, Tool):
            raise TypeError(f"必须是Tool类型，得到: {type(tool)}")
        
        self._tools[tool.name] = tool
        print(f"[TOOL] 工具 '{tool.name}' 已注册")
    
    def register_function(
        self, 
        name: str, 
        description: str, 
        func: Callable
    ) -> None:
        """
        注册函数作为工具
        
        Args:
            name: 工具名称
            description: 工具描述
            func: 可调用函数
        """
        self._functions[name] = {
            "description": description,
            "func": func
        }
        print(f"[TOOL] 函数工具 '{name}' 已注册")
    
    def unregister(self, name: str) -> bool:
        """
        注销工具
        
        Args:
            name: 工具名称
            
        Returns:
            是否成功注销
        """
        if name in self._tools:
            del self._tools[name]
            return True
        if name in self._functions:
            del self._functions[name]
            return True
        return False
    
    def get_tool(self, name: str) -> Optional[Tool]:
        """
        获取工具实例
        
        Args:
            name: 工具名称
            
        Returns:
            工具实例，如果不存在返回None
        """
        return self._tools.get(name)
    
    def list_tools(self) -> list:
        """
        列出所有已注册的工具名称
        
        Returns:
            工具名称列表
        """
        all_tools = list(self._tools.keys()) + list(self._functions.keys())
        return all_tools
    
    def get_tools_description(self) -> str:
        """
        获取所有工具的描述信息
        
        Returns:
            格式化的工具描述字符串
        """
        if not self._tools and not self._functions:
            return "暂无可用工具"
        
        descriptions = []
        
        # 添加工具类描述
        for name, tool in self._tools.items():
            descriptions.append(f"- {name}: {tool.description}")
        
        # 添加函数工具描述
        for name, info in self._functions.items():
            descriptions.append(f"- {name}: {info['description']}")
        
        return "\n".join(descriptions)
    
    def execute_tool(self, tool_name: str, *args, **kwargs) -> Any:
        """
        执行指定的工具
        
        Args:
            tool_name: 工具名称
            *args: 位置参数
            **kwargs: 关键字参数
            
        Returns:
            工具执行结果
            
        Raises:
            ValueError: 工具不存在
        """
        # 检查是否是工具类
        if tool_name in self._tools:
            tool = self._tools[tool_name]
            return tool.run(*args, **kwargs)
        
        # 检查是否是函数工具
        if tool_name in self._functions:
            func = self._functions[tool_name]["func"]
            return func(*args, **kwargs)
        
        raise ValueError(f"工具 '{tool_name}' 不存在")
    
    def __len__(self) -> int:
        """返回注册的工具数量"""
        return len(self._tools) + len(self._functions)
    
    def __contains__(self, name: str) -> bool:
        """检查工具是否存在"""
        return name in self._tools or name in self._functions

File: src/tools/registry.ts
/**
 * 工具注册表 - 管理和执行工具
 */

import { Tool, ToolParameters, FunctionSchema } from './base.js';
import { Logger, silentLogger } from '../core/logger.js';

/**
 * 函数工具信息
 */
interface FunctionToolInfo {
  description: string;
  func: (input: string) => Promise<string> | string;
}

/**
 * 工具注册表配置
 */
export interface ToolRegistryOptions {
  logger?: Logger;
}

/**
 * 工具注册表
 *
 * 提供工具的注册、管理和执行功能。
 * 支持两种工具注册方式：
 * 1. Tool 对象注册（推荐）
 * 2. 函数直接注册（简便）
 */
export class ToolRegistry {
  private tools: Map<string, Tool> = new Map();
  private functions: Map<string, FunctionToolInfo> = new Map();
  private readonly logger: Logger;

  constructor(options: ToolRegistryOptions = {}) {
    this.logger = options.logger ?? silentLogger;
  }

  /**
   * 注册 Tool 对象
   */
  registerTool(tool: Tool, autoExpand = true): void {
    // 检查工具是否可展开
    if (autoExpand && tool.expandable) {
      const expandedTools = tool.getExpandedTools();
      if (expandedTools && expandedTools.length > 0) {
        for (const subTool of expandedTools) {
          if (this.tools.has(subTool.name)) {
            this.logger.warn(`工具 '${subTool.name}' 已存在，将被覆盖`);
          }
          this.tools.set(subTool.name, subTool);
        }
        this.logger.debug(`工具 '${tool.name}' 已展开为 ${expandedTools.length} 个独立工具`);
        return;
      }
    }

    // 普通工具或不展开的工具
    if (this.tools.has(tool.name)) {
      this.logger.warn(`工具 '${tool.name}' 已存在，将被覆盖`);
    }

    this.tools.set(tool.name, tool);
    this.logger.debug(`工具 '${tool.name}' 已注册`);
  }

  /**
   * 直接注册函数作为工具
   */
  registerFunction(
    name: string,
    description: string,
    func: (input: string) => Promise<string> | string
  ): void {
    if (this.functions.has(name)) {
      this.logger.warn(`工具 '${name}' 已存在，将被覆盖`);
    }

    this.functions.set(name, { description, func });
    this.logger.debug(`工具 '${name}' 已注册`);
  }

  /**
   * 注销工具
   */
  unregister(name: string): void {
    if (this.tools.has(name)) {
      this.tools.delete(name);
      this.logger.debug(`工具 '${name}' 已注销`);
    } else if (this.functions.has(name)) {
      this.functions.delete(name);
      this.logger.debug(`工具 '${name}' 已注销`);
    } else {
      this.logger.warn(`工具 '${name}' 不存在`);
    }
  }

  /**
   * 获取 Tool 对象
   */
  getTool(name: string): Tool | undefined {
    return this.tools.get(name);
  }

  /**
   * 获取函数工具
   */
  getFunction(name: string): ((input: string) => Promise<string> | string) | undefined {
    return this.functions.get(name)?.func;
  }

  /**
   * 执行工具
   */
  async executeTool(name: string, input: string | ToolParameters): Promise<string> {
    // 优先查找 Tool 对象
    const tool = this.tools.get(name);
    if (tool) {
      try {
        const params = typeof input === 'string' ? { input } : input;
        const result = await tool.run(params);
        return result;
      } catch (error) {
        const message = error instanceof Error ? error.message : String(error);
        return `错误：执行工具 '${name}' 时发生异常: ${message}`;
      }
    }

    // 查找函数工具
    const funcInfo = this.functions.get(name);
    if (funcInfo) {
      try {
        const inputStr = typeof input === 'string' ? input : JSON.stringify(input);
        const result = await funcInfo.func(inputStr);
        return result;
      } catch (error) {
        const message = error instanceof Error ? error.message : String(error);
        return `错误：执行工具 '${name}' 时发生异常: ${message}`;
      }
    }

    return `错误：未找到名为 '${name}' 的工具。`;
  }

  /**
   * 获取所有可用工具的格式化描述
   */
  getToolsDescription(): string {
    const descriptions: string[] = [];

    // Tool 对象描述
    for (const tool of this.tools.values()) {
      descriptions.push(`- ${tool.name}: ${tool.description}`);
    }

    // 函数工具描述
    for (const [name, info] of this.functions) {
      descriptions.push(`- ${name}: ${info.description}`);
    }

    return descriptions.length > 0 ? descriptions.join('\n') : '暂无可用工具';
  }

  /**
   * 列出所有工具名称
   */
  listTools(): string[] {
    return [...this.tools.keys(), ...this.functions.keys()];
  }

  /**
   * 获取所有 Tool 对象
   */
  getAllTools(): Tool[] {
    return [...this.tools.values()];
  }

  /**
   * 获取所有工具的 OpenAI Schema
   */
  getAllToolSchemas(): FunctionSchema[] {
    const schemas: FunctionSchema[] = [];

    // Tool 对象的 schema
    for (const tool of this.tools.values()) {
      schemas.push(tool.toOpenAISchema());
    }

    // 函数工具的 schema
    for (const [name, info] of this.functions) {
      schemas.push({
        type: 'function',
        function: {
          name,
          description: info.description,
          parameters: {
            type: 'object',
            properties: {
              input: {
                type: 'string',
                description: '输入文本',
              },
            },
            required: ['input'],
          },
        },
      });
    }

    return schemas;
  }

  /**
   * 清空所有工具
   */
  clear(): void {
    this.tools.clear();
    this.functions.clear();
    this.logger.debug('所有工具已清空');
  }

  /**
   * 获取工具数量
   */
  get size(): number {
    return this.tools.size + this.functions.size;
  }
}

/**
 * 全局工具注册表
 */
export const globalRegistry = new ToolRegistry();
